open_city_profile/middleware.py

```diff
@@ -0,0 +1,26 @@
+from django.conf import settings
+from django.core.exceptions import MiddlewareNotUsed
+from helusers.oidc import RequestJWTAuthentication
+
+
+class JWTAuthentication:
```

Comment:
Is there a good reason for this being a middleware and not a proper authentication backend?

Reply:
The Django authentication backends can only report authenticated/not authenticated, without any reasons why (not authenticated). We want to return some information about the reasons.

Django also mixes authentication with authorization (as if both could be performed at the same point in the request handling), but with GraphQL requests that doesn't really work.

It's possible to make a GraphQL query, where e.g. one part could need an authenticated user and another one not. In such case we always return data for the part that doesn't require authentication, and check the authentication status for the part requiring it and return either data or error (with an appropriate reason) for that part. Authentication and authorization happen in different stages with GraphQL and authorization can even happen multiple times, with different criteria.